libraries/Microsoft.Bot.Builder/TokenResolver.cs

```diff
@@ -163,5 +150,78 @@ private static Activity CreateTokenResponse(ConversationReference relatesTo, str
 
             return tokenResponse;
         }
+
+        private class PollingParams
+        {
+            public bool SentToken { get; set; }
+
+            public bool ShouldEndPolling { get; set; }
+
+            public TimeSpan PollingTimeout { get; set; }
+
+            public TimeSpan PollingInterval { get; set; }
+
+            public string ConnectionName { get; set; }
+        }
+
+        private class PollingHelper
+        {
+            public TimeSpan DefaultPollingInterval { get; set; }
+
+            public TimeSpan DefaultPollingTimeout { get; set; }
+
+            public CancellationToken CancellationToken { get; set; }
+
+            public BotAdapter Adapter { get; set; }
+
+            public ILogger Logger { get; set; }
+
+            public Activity Activity { get; set; }
+
+            public ClaimsIdentity Identity { get; set; }
+
+            public BotCallbackHandler Callback { get; set; }
+
+            public async Task PollForTokenAsync(PollingParams pollingParams)
+            {
+                BotCallbackHandler continueCallback = async (context, ctoken) =>
+                {
+                    // TODO: Should be using OAuthClient
+                    //var oauthClient = context.TurnState.Get<OAuthClient>();
+                    //var tokenResponse = await oauthClient.UserToken.GetTokenAsync(Activity.From.Id, pollingParams.ConnectionName, Activity.ChannelId, null, ctoken).ConfigureAwait(false);
```

Comment:
These comments are about using the raw [OAuthClient](https://github.com/microsoft/botbuilder-dotnet/blob/152ad50ac31ac45c663622298e1cc6036346c3f6/libraries/Microsoft.Bot.Connector/OAuthClient.cs#L19) which makes the implementation inherently tied to the Bot Framework User Token Service.

Ideally this would be tied to the proposed `CloudUserTokenClient` in https://github.com/microsoft/botbuilder-dotnet/pull/5062.

Reply:
Yes this should use the UserTokenClient abstract class. We could even split that abstraction and pull this method into its own abstract class if we wanted to shim BotFrameworkAdapter.

UserTokenClient is very close to the raw OAuthClient except it captures some protocol payload the underlying client doesn't model. The implementation also abstracts certs - the need to provide that on every method call in teh ExtendedUserToken interface was one of its main problems.

Reply:
But I think we could take this PR as a nice step in the right direction. @stevengum thoughts?

Reply:
I agree! I think taking the PR was a step in the right direction.